frontend/model/contracts/group.js

```diff
@@ -339,6 +360,52 @@ DefineContract({
         archiveProposal(state, data.proposalHash)
       }
     },
+    'gi.contracts/group/removeMember': {
+      validate: (data, { state, getters }) => {
+        objectOf({
+          member: string,
+          memberId: string,
+          groupId: string,
+          // In case it happens in a big group (by proposal)
+          // we need to validate the associated proposal.
+          proposalHash: optional(string),
+          proposalPayload: optional(objectOf({
+            secret: string // NOTE: simulate the OP_KEY_* stuff for now
+          }))
+        })(data)
+
+        const memberToRemove = data.member
+        const membersCount = getters.groupMembersCount
+
+        if (!state.profiles[memberToRemove]) {
+          throw new TypeError('Not part of the group.')
+        }
+        if (membersCount === 1) {
+          throw new TypeError('Cannot remove yourself.')
+        }
+        if (membersCount >= 3) {
+          const { payload } = state.proposals[data.proposalHash] || {}
+          if (!payload || payload.secret !== data.proposalPayload.secret) {
+            throw new TypeError('Invalid associated proposal.')
+          }
+        }
```

Comment:
I do not get this invalidation function or why there's a secret.

I remember us reaching the conclusion on Figma that check for `group/removeMember` would be if the group creator was the one sending the message, and only in the case that voting isn't required (i.e. less than 3 members). That's not what this function is doing.

Reply:
Yes, we agreed that on Figma, but we still need a validation somehow when the group has >= 3 members, right? So we can prevent someone from calling this method hardcoded and "bypassing the proposal".

I thought to do the same as we do with invite a member. We create the secret at the moment the proposal is approved and send it through `proposals.js`

Reply:
> We create the secret at the moment the proposal is approved and send it through proposals.js. Does it make sense?

OK, the issue is that this validation function is only ever called as a result of a `'gi.contracts/group/removeMember'` message being created and sent to the group only. This does not happen anywhere in this PR. When the proposal passes, we bypass this validation code and call `'gi.contracts/group/removeMember/process'` and `'gi.contracts/group/removeMember/process/sideEffect'` directly from within `proposals.js`, and that does not happen as a result of a `gi.contracts/group/removeMember` message, but as a result of a `'gi.contracts/group/proposalVote'` message.

So this validation code doesn't get run at all anywhere in this PR.

And maybe that's one issue to consider. In other words, whether `proposal.js` should be modified to call the validation code as well. Maybe that should be a separate issue, because `Contract.js` does not expose the validate function via SBP (there are no `group/invite/validate` or `group/removeMember/validate` selectors).

However, that isn't even the biggest issue I'm noticing. The biggest issue is that `removeMember` is not really designed to be called from a proposal. It is designed to be called by the admin to remove a member without a proposal. So with the way this is implemented, it would in fact throw an exception, because `payload` would be undefined.

So these lines 386-389 should be removed, and instead the check should be whether `getters.groupSettings.groupCreator === meta.username`.

Reply:
> It is designed to be called by the admin to remove a member without a proposal. 

Hum... no, it's designed to be called by user1 (creator) or user2 as far as the group has only 2 members. ATM user2 can remove user1 if they want to. (in a group with 2 members). The same happens in a group with 3 members but through a proposal.

So that's why the validation is `membersCount >= 3` and check for the proposal secret to make sure it was triggered by a valid proposal.

Reply:
> Hum... no, it's designed to be called by user1 (creator) or user2 as far as the group has only 2 members. ATM user2 can remove user1 if they want to. (in a group with 2 members).

Yes, so on Figma we agreed that we would treat the creator as the admin in this case.

This code needs to be added here.

The proposal secret should not be checked here, because of what I wrote above. The proposal secret is never validated with the way this PR is written. That if statement is not only doing nothing, it will cause the code to break should we ever implement a UI for the creator to remove members when the member count is less than 3 members.

Reply:
> We agreed that we would treat the creator as the admin in this case.

I changed the code to do that, but what's the problem of user2 being able to remove user1? The validation of a group having less than 3 members isn't enough?

I also removed the proposal secret, since it's not needed after all.

Reply:
> I changed the code to do that, but what's the problem of user2 being able to remove user1? The validation of a group having less than 3 members isn't enough?

The issue is that it would be a poor user experience for users creating groups if they created a group and invited someone to join the group, and could be immediately removed by the person they invited. As unlikely as that would be to ever happen, we still shouldn't allow it.

> So where should it be done?

It shouldn't be done at all, and all the code related to the proposal secret should be removed. :)

> The way I see it, it works fine here because this verification is indeed executed when a proposal is approved. It's called at proposals.js L39 by removeMember/process

I just had a look at the code in `proposals.js` in this PR and it does not result in any calls to the validate function. The proposal secret isn't needed in this PR.

The check here should simply throw an exception in the else. Do not worry, it will not get triggered, because the proposal code doesn't result in a call to this validation stuff anyway.

Add a `// TODO: check whether meta.username has required admin permissions` comment in the else as well, and we can deal with that in the future

Reply:
> I just had a look at the code in proposals.js in this PR and it does not result in any calls to the validate function. The proposal secret isn't needed in this PR.

Well... that's not true: I just saw it by myself. On the `else` I added a warn:
```
console.warn('This secret for this proposal is:', payload.secret)
```

And after the proposal gets approved, it is logged into the console, it is logged once per member:

![image](https://user-images.githubusercontent.com/14869087/75488067-91474900-59a7-11ea-9d18-9269ed401434.png)

So... that's why I think the secret is needed.

Reply:
Where did you add that console.warn? Can you push your changes? It doesn't make sense to me why the validate function would be called for a remove member proposal, since the validate function for `'gi.contracts/group/removeMember'` is only ever called when a `'gi.contracts/group/removeMember'` action is sent directly, and that only happens when there's no proposal.

Reply:
*Edit to reflect how wrong I am. OK, now let me re-process this and re-think if it's still needed...

Reply:
right here: 
![image](https://user-images.githubusercontent.com/14869087/75488981-44647200-59a9-11ea-9061-810e2598deb8.png)

Reply:
OK. Then here's what needs to be done: the condition should throw in two cases:

1. If there is a proposal, then check if there is a proposal secret and throw if they don't match. I think what you're currently doing here is throwing if there is no proposal too... but we need a slightly different error in that case.
2. If there is no corresponding proposal, then we need to throw a different error, which says that `'admin credentials needed and not implemented yet'`, or something to that effect

Reply:
done, please see again